kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/bitwise-simplification.k

```diff
@@ -64,10 +81,12 @@ module BITWISE-SIMPLIFICATION [symbolic]
        andBool 0 <=Int Y andBool Y <Int 2 ^Int ( 256 -Int Z )
       [simplification, concrete(X, Z), comm, preserves-definedness]
 
-    // Generalization of: maxUIntXXX &Int (Y |Int (notMaxUIntXXX &Int T)) => maxUIntXXX &Int Y
+    // Generalization of:    maxUIntXXX &Int (Y |Int (notMaxUIntXXX &Int T)) =>    maxUIntXXX &Int Y
+    //                and notMaxUIntXXX &Int (Y |Int (   maxUIntXXX &Int T)) => notMaxUIntXXX &Int Y
     rule X &Int (Y |Int (Z &Int T)) => X &Int Y
       requires 0 <=Int X andBool 0 <=Int Y andBool 0 <=Int Z andBool 0 <=Int T
-       andBool X +Int 1 ==Int 2 ^Int log2Int(X +Int 1)
+       andBool (      X +Int 1 ==Int 2 ^Int log2Int (      X +Int 1 ) orBool // X is a maxUInt
+                 pow256 -Int X ==Int 2 ^Int log2Int ( pow256 -Int X ) )      // X is a notMaxUInt
```

Comment:
This rule, interestingly, works for both sets of lemmas. All we need is that `orBool`, which is guaranteed to be always concrete, so there should be no issues. The very specific shape of the LHS term means that the rule won't be tried often.